### DISCUSSION TITLE
We would like to initialize an object at declaration.

### DISCUSSION THREAD

**Discussion by PTKu** [simatic-ax/.discussions#13] [Ideas]
We would like to initialize an object at declaration. 

Something along the lines of https://content.helpme-codesys.com/en/CODESYS%20Development%20System/_cds_method_fb_init_fb_reinit.html.

~~~Smalltalk
CLASS PUBLIC ClassWithSomeKindOfConstructor
     VAR PRIVATE
        _somePrivateMember : INT;
     END_VAR

     // This would act as a constructor
     METHOD PUBLIC ClassInit
         VAR_INPUT
             InitPrivateMemberWithValue : INT;
         END_VAR
         _somePrivateMember := InitPrivateMemberWithValue + 10;
     END_VAR
END_CLASS
~~~

~~~Smalltalk
.
.
.
.
VAR
    myClass : ClassWithSomeKindOfConstructor(10);
END_VAR
.
.
.
.
~~~

Reason:

We would need to have control over the object construction at initialization. At present, it is possible to make assignments at declaration. We would need to have the possibility to execute a code to provide additional logic to initialize an object. This would allow initializing otherwise inaccessible members but also execute initialization logic.

**Comment by sjuergen** [comment]
Hi @PTKu, thanks for that topic. I'm not sure whether your Codesys example follows the ICE61131-3. I haven't found something in that direction (@theBadT could you provide any insights there?).

But for classes, you could do in an applicative way:

``` 
CLASS ABSTRACT Base
    VAR PRIVATE
        isInitialized : BOOL;
    END_VAR
    METHOD PUBLIC RunCyclic
        IF (isInitialized) THEN
            THIS.Init();
            isInitialized := TRUE;
        END_IF;
        THIS.RunCyclicUseCode();
    END_METHOD
    METHOD ABSTRACT RunCyclicUseCode
    END_METHOD
    METHOD ABSTRACT Init
    END_METHOD
END_CLASS
```

Inherit your class from the Base class and implement the Init() and RunCyclicUseCode() method.
```
CLASS ConcreteClass
    EXTENDS Base
    VAR
        a : INT;
    END_VAR
    METHOD PROTECTED OVERRIDE RunCyclicUseCode
        // Implement your user code here
        a := a + 1;
    END_METHOD
    // Your "applicative constructor without parameters"
    METHOD PROTECTED OVERRIDE Init
        // Do your initialization here
        a := 10;
    END_METHOD
END_CLASS
``` 

For sure there are some limitations (e.g. the init() method does not allow different interfaces. But basically that could be a workaround.

**Comment by PTKu** [comment]
Thanks, @sjuergen. You are right. We can initialize the code in the call tree of a program. However, the idea with the initialization at construction is that it would enforce the call of the init method (not initialized would produce compile time error). **This would prevent declaring a variable/field that would not be properly initialized and thus avoid possible errors at runtime**. It proved extremely useful for us in the past.

You're also right that there is no such thing in the current IEC61131-3 specification. As a side note to IEC. They are releasing the standard at a cadence of about ten years (the next release is delayed by about three years), which is way too much for any language to progress in a meaningful way. I would not mind if you added something outside the standard;).

It is really about making sure the initialization was called. It would not be necessary to add 'init' method to the language; maybe it would suffice to check that the 'init; was called before the instance was used. It could be enforced by a compiler analyzer. I am unsure about the implementation, as this would need to know the call path to determine that the instance was initialized, and it may get tricky. Probably the more straightforward way would be to require the initialization in the declaration.

**Reply by sjuergen** [reply]
@PTKu I agree, my applicative approach is not perfect. It assumes, that you call the RunCyclic() at minimum once (could also be done in the Startup task) before calling other methods of that class. For sure, you've to care about the call.

Since the both methods Init() and RunCyclicUseCode() are abstract methods, you're enforced to implement both methods which are called in the RunCyclic() method.

Regarding the implementation in the ST language, I can't make any statement. But as I understood it correctly, it's implemented in CodeSys. Isn't it? Do you know other ST solutions which has a support of constructors?

**Comment by PTKu** [comment]
@sjuergen, as far as I know, FB_init is implemented in CoDeSys and the derivatives like Beckhoff's TwinCAT platform. I know that there is a handful of people that use this kind of initialization on those platforms, so it might be worth considering something like that in AX as well.
It would also come in handy in our [application framework](https://github.com/PTKu/ix.framework). For the moment, we just have a method `Initialize` that is required to be called explicitly at some appropriate place in the program. However, had we had this option, it would simplify the use.
